src/lexer/lexer.cc

```diff
@@ -0,0 +1,35 @@
+#include "lexer/lexer.h"
+
+#include <fstream>
+#include <sstream>
+
+namespace lexer {
+
+namespace {
+
+std::istream* get_stream_from_source(const std::string& source, Lexer::SourceTag tag) {
+  switch (tag) {
+    case Lexer::SourceTag::FILE:
+      return new std::ifstream(source);
+    case Lexer::SourceTag::STRING:
+      return new std::stringstream(source);
+    default:
+      return nullptr;
```

Comment:
I'm not a huge fan of silently returning a nullptr in case of error. It would be nice to either throw an exception (this case really shouldn't happen) and/or logging something.